In [1]:
import urllib3
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

In [2]:
import requests
from bs4 import BeautifulSoup

class BaseLoader:
    def __init__(self, urls):
        self.urls = urls
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'
        }

    def load(self):
        references = []
        for url in self.urls:
            response = requests.get(url, headers=self.headers, verify=False)
            soup = BeautifulSoup(response.text, 'html.parser')
            references.append(soup.get_text())
        
        return references


In [3]:
with open("hypos.txt") as f:
    hypos = [line.strip() for line in f]

with open("refs.txt") as f:
    refs = [line.strip() for line in f]

In [4]:
refs

['https://baoantelecom.com/vi/tin/dien-tu-vien-thong',
 'https://www.hotcourses.vn/study-abroad-info/study-guides/nganh-dien-tu-vien-thong/',
 'https://blog.topcv.vn/co-nen-hoc-nganh-dien-tu-vien-thong/']

In [5]:
loader = BaseLoader(refs)
refs= loader.load()

In [6]:
len(refs)

3

In [7]:
refs

['\n\n\n\n\n\n\n\n\nTìm hiểu kiến thức học và cơ hội việc làm ngành Điện tử viễn thông\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n \xa0\n                    Hồ Chí Minh: 0918.325.086 \xa0 \xa0 - \xa0 \xa0\n                    Hà Nội:  0911.527.086 \xa0 \xa0 - \xa0 \xa0\n                    Hotline: 0918.132.458 \n\n\n\n\n\nNgôn ngữ :\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nTrang chủ\n\nGiới thiệu\n\n\nGIỚI THIỆU CHUNG\n\n\nĐỐI TÁC\n\n\nTuyển dụng\n\n\n\n\nSản phẩm\n\n\n\n\n\n\n\n\n\n                                                    Hội nghị truyền hình\n                                                \n\nAVer\nPolycom\nCisco\nLifeSize\nAVaya\nSony\nSurf\nCrestron\nTely\n\n\n\n\n\n\n\n\n\n\n\n\n\n                                                    Camera hội nghị\n                                                \n\nCamera Hội Nghị AVer\nCamera hội nghị IVC\nCamera hội nghị MAXHUB\n\n\n\n\n\n\n\n\n\n\n\n\n\n                

In [8]:
from unidecode import unidecode
import re

def normalize_text(text):
    """
    Normalize Vietnamese text by converting it to its ASCII representation and removing accents.
    
    Args:
    text (str): Input Vietnamese text to be normalized.
    
    Returns:
    str: Normalized text with accents removed.
    """
    normalized_text = unidecode(text)
    normalized_text = re.sub(r'\s+', ' ', normalized_text).strip()
    return normalized_text


In [9]:
hypos = [' '.join(normalize_text(hypo) for hypo in hypos)]

In [10]:
hypos

['# Co Hoi Nghe Nghiep Cho Sinh Vien Nganh Dien Tu - Vien Thong Cua Dai Hoc Bach Khoa Ha Noi Sau Khi Tot Nghiep  Nganh Dien tu - Vien thong da va dang tro thanh mot trong nhung nganh hoc quan trong va khong the thieu trong boi canh cuoc cach mang cong nghe 4.0 dang dien ra manh me tren toan cau. Doi voi sinh vien Dai hoc Bach khoa Ha Noi, mot trong nhung truong dai hoc hang dau Viet Nam, co hoi nghe nghiep sau khi tot nghiep nganh nay la vo cung rong mo va da dang.  ## Tong Quan ve Nganh Dien tu - Vien Thong  Nganh Dien tu - Vien thong bao gom viec nghien cuu, thiet ke, phat trien va ung dung cac cong nghe lien quan den dien tu va vien thong. Sinh vien nganh nay duoc dao tao de tro thanh nhung ky su co kha nang lam viec trong moi truong cong nghe cao, dap ung nhu cau ngay cang tang ve viec thiet ke va van hanh cac he thong thong tin va truyen thong hien dai (TopCV, n.d.).  ## Co Hoi Nghe Nghiep Sau Khi Tot Nghiep  ### Cac Vi Tri Cong Viec Tiem Nang  Sinh vien tot nghiep nganh Dien tu -

In [11]:
refs = [normalize_text(ref) for ref in refs]

In [12]:
refs

['Tim hieu kien thuc hoc va co hoi viec lam nganh Dien tu vien thong Ho Chi Minh: 0918.325.086 - Ha Noi: 0911.527.086 - Hotline: 0918.132.458 Ngon ngu : Trang chu Gioi thieu GIOI THIEU CHUNG DOI TAC Tuyen dung San pham Hoi nghi truyen hinh AVer Polycom Cisco LifeSize AVaya Sony Surf Crestron Tely Camera hoi nghi Camera Hoi Nghi AVer Camera hoi nghi IVC Camera hoi nghi MAXHUB Man hinh hien thi Man hinh ghep Man hinh tuong tac PHU KIEN HOI NGHI TU SAC He thong am thanh phong hop LOA AMPLY MICROPHONE MIXER He thong dieu khien thong minh Thiet bi tuong tac Thiet bi chia se khong day Thiet bi dieu khien phong hop Du an Hoi nghi truyen hinh Man hinh ghep Giai phap Hoi nghi truyen hinh Man hinh ghep Man hinh tuong tac Dich vu LAP DAT CHO THUE SUA CHUA THIET BI Tin tuc hoi nghi truyen hinh Man hinh ghep Man hinh tuong tac Lien he Tin tuc Trang chu Tim hieu kien thuc hoc va co hoi viec lam nganh Dien tu vien thong Tim hieu kien thuc hoc va co hoi viec lam nganh Dien tu vien thong Vo Kim Ngan 11

In [13]:
len(hypos), len(refs)

(1, 3)

### Using BERTScore for Evaluation

In [14]:
from bert_score import score

/media/sanslab/Data/stevehoang/rag_research/vigpt-researcher/venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [15]:
# Hide the loading messages
import logging
import transformers
transformers.tokenization_utils.logger.setLevel(logging.ERROR)
transformers.configuration_utils.logger.setLevel(logging.ERROR)
transformers.modeling_utils.logger.setLevel(logging.ERROR)

In [16]:
def bert_score(hypos, refs, lang="vi"):
    bert_scores = []
    for ref in refs:
        scores = score(hypos, [ref], lang=lang, verbose=False)
        bert_scores.append(scores)
    return bert_scores

In [17]:
bert_scores = bert_score(hypos, refs)

In [18]:
import torch

precision = torch.mean(torch.stack([bert_score[0] for bert_score in bert_scores]))
recall = torch.mean(torch.stack([bert_score[1] for bert_score in bert_scores]))
f1 = torch.mean(torch.stack([bert_score[2] for bert_score in bert_scores]))

In [19]:
print("BERT score:")
print(f"P: {precision.item():.2f}, R: {recall.item():.2f}, F1: {f1.item():.2f}")

BERT score:
P: 0.68, R: 0.69, F1: 0.68


### Using ROUGE score for Evaluation

In [20]:
from rouge_score import rouge_scorer

def rouge_score(hypos, refs):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2'], use_stemmer=True)
    scores = [scorer.score(ref, hypo) for ref, hypo in zip(refs, hypos)]
    return scores

rouge = rouge_score(hypos, refs)

In [21]:
for i, score in enumerate(rouge):
    print("ROUGE scores:")
    print(f"ROUGE-1: P = {score['rouge1'].precision:.2f}, R = {score['rouge1'].recall:.2f}, F1 = {score['rouge1'].fmeasure:.2f}")
    print(f"ROUGE-2: P = {score['rouge2'].precision:.2f}, R = {score['rouge2'].recall:.2f}, F1 = {score['rouge2'].fmeasure:.2f}")
    print()

ROUGE scores:
ROUGE-1: P = 0.83, R = 0.39, F1 = 0.53
ROUGE-2: P = 0.42, R = 0.20, F1 = 0.27



### RESULTS

| METRIC   | Precision | Recall | F1 |
|---------|----------|-------|-------|
| BERTScore|0.69, 0.67, 0.68| 0.68, 0.66, 0.69|0.69, 0.66, 0.68|
| ROUGE-1|0.86, 0.88, 0.83|0.31, 0.29, 0.39|0.46, 0.44, 0.53|
| ROUGE-2|0.39, 0.40, 0.42|0.14, 0.13, 0.20|0.21, 0.20, 0.27|

* BERTScore Average: Precision = ?, Recall = ?, F1 = ?
* ROUGE-1 Average: Precision = ?, Recall = ?, F1 = ?
* ROUGE-2 Average: Precision = ?, Recall = ?, F1 = ?

### Review
1. Summarization is a complex task, even advanced models can struggle to accurately capture all the important information from a source text.

2. BERTScore or ROUGE that these metrics focus on things like including keywords from the source text, so that do not perfectly reflect human judgement of a good summary.

3. The response of chatbot from many different sources.


### HUMAN EVALUATION IS THE BEST WAY TO EVALUATE THE QUALITY OF A SUMMARY